client/rest/src/plugins/metadata/metal.js

```diff
@@ -0,0 +1,214 @@
+	 * - scopedMetadataKey: uint64
+	 * - chunkPayload: Buffer
+	 */
+	extractChunk(b) {
+		const isValidSize = 12 < b.length && 1024 >= b.length;
+		if (!isValidSize)
+			throw new Error(`Invalid metadata value size ${b.length}`);
+
+		const header = b.subarray(0, 1);
+		const magic = header[0] & 0x80;
+		const text = Boolean(header[0] & 0x40);
+
+		const scopedMetadataKey = catapult.utils.uint64.fromBytes(new Uint8Array(b.subarray(4, 12)).reverse());
+		const chunkPayload = b.subarray(12);
+
+		return {
+			magic, text, scopedMetadataKey, chunkPayload
+		};
+	},
+
+	/**
+	 * Separate by null if text is present or not.
+	 * @param {object} chunkData chunkPayload and has text or not
+	 * - chunkPayload: Buffer
+	 * - text: boolean
+	 * @returns {object} return chunkPayload and chunkText.
+	 * - chunkPayload: Uint8Array
+	 * - chunkText: Uint8Array
+	 */
+	splitChunkPayloadAndText(chunkData) {
+		if (!chunkData.text) {
+			// No text in the chunk
+			return {
+				chunkPayload: chunkData.chunkPayload,
+				chunkText: undefined
+			};
+		}
+
+		// Extract text section until null char is encountered.
+		const textBytes = [];
+		for (let i = 0; i < chunkData.chunkPayload.length && chunkData.chunkPayload[i]; i++)
+			textBytes.push(chunkData.chunkPayload[i]);
+
+		return {
+			chunkPayload: new Uint8Array(chunkData.chunkPayload.subarray(textBytes.length + 1)),
+			chunkText: new Uint8Array(textBytes)
+		};
+	},
+
+	/**
+	 * Decode binary data from chunks.
+	 * @param {Uint64} fitstKey ScopedMetadataKey of the beginning chunk.
+	 * @param {Array} chunks Chunk containing data to be decoded.
+	 * @returns {Buffer} Decoded binary data.
+	 */
+	decode(fitstKey, chunks) {
+		let scopedMetadataKey = fitstKey;
+		let magic;
+		let decodedPayload = Buffer.from([]);
+		let decodedText = Buffer.from([]);
+		const findChunk = c => 0 === catapult.utils.uint64.compare(c.key, scopedMetadataKey);
+		do {
+			const chunk = chunks.find(findChunk);
+			if (!chunk)
+				throw new Error(`Error: The chunk ${scopedMetadataKey} is missing`);
+
+			const checksum = this.generateMetadataKey(chunk.value);
+			const isValidChecksum = 0 === catapult.utils.uint64.compare(checksum, scopedMetadataKey);
+			if (!isValidChecksum)
+				throw new Error(`Error: The chunk ${scopedMetadataKey} is broken (calculated=${checksum})`);
+
+			const chunkData = this.extractChunk(chunk.value);
+			({ magic, scopedMetadataKey } = chunkData);
+			const { chunkPayload, chunkText } = this.splitChunkPayloadAndText(chunkData);
+			if (chunkPayload.length)
+				decodedPayload = Buffer.concat([decodedPayload, chunkPayload]);
+
+			if (chunkText?.length)
+				decodedText = Buffer.concat([decodedText, chunkText]);
+		} while (0x80 !== (magic & 0x80));
+		return {
+			payload: decodedPayload,
+			text: decodedText.length ? decodedText.toString('utf-8') : undefined
+		};
+	},
+
+	/**
+	 * Get metadata entry by compositehash.
+	 * I can't throw exceptions in db tests, so I made it for unit tests.
```

Comment:
at first, `binDataByMetalId` in `MetadataDb.js` is below
```js
async binDataByMetalId(metalId) {
const compositeHashes = [routeUtils.parseArgument(metal.restoreMetadataHash(metalId), 'compositeHash', 'hash256')];
const metadatasByCompositeHash = await this.metadatasByCompositeHash(compositeHashes);
if (0 === metadatasByCompositeHash.length)
	throw Error('could not get first chunk, it may mistake the metal ID.');
const metadataEntry = metadatasByCompositeHash[0].metadataEntry;
```
however, i cannot test throw error with `runMetadataDbTest`
like this
```js
it.only('cannot decode', () =>
	// Act + Assert:
	runMetadataDbTest(
		dbMosaicMetadata,
		db => db.binDataByMetalId('Fe4YG12YcUzgATsZexNAhLyfbxogSaLX7dhoHMvCqgnPac'),
		d => expect(d.payload).to.throw(`could not get first chunk, it may mistake the metal ID.`)			
	));
```
so, i put the function into `metal.js` for testing.

Reply:
i think something like this should work for you:
```js
	runMetadataDbTest(
		dbMosaicMetadata,
		db => expect(() => db.binDataByMetalId('Fe4YG12YcUzgATsZexNAhLyfbxogSaLX7dhoHMvCqgnPac')).to.throw('...'),
		() => {} // assertion done as part of (failed) issueDbCommand
	);
```

Reply:
I tried to test several patterns and this worked. Is this still good?
```js
it('cannot decodes not getting first chunk', () =>
	// Act + Assert:
	runMetadataDbTest(
		dbMetadata(),
		async db => {
			try {
				await db.binDataByMetalId('Fe4YG12YcUzgATsZexNAhLyfbxogSaLX7dhoHMvCqgnPao');
			} catch (error) {
				expect(error.message).to.equal('could not get first chunk, it may mistake the metal ID: Fe4YG12YcUzgATsZexNAhLyfbxogSaLX7dhoHMvCqgnPao');
			}
		},
		() => {} // assertion done as part of (failed) issueDbCommand
	)
);
```

Reply:
the problem with that structure is that if exception is NOT thrown, test will still pass

try this:
```js
	runMetadataDbTest(
		dbMosaicMetadata,
		db => expect(() => { await db.binDataByMetalId('Fe4YG12YcUzgATsZexNAhLyfbxogSaLX7dhoHMvCqgnPac')).to.throw('...'); },
		() => {} // assertion done as part of (failed) issueDbCommand
	);
```

Reply:
this you thought me worked!!(little bit changed
```js
it('cannot decode not getting first chunk', () => {
	// Arrange:
	const metalId = 'Fe4YG12YcUzgATsZexNAhLyfbxogSaLX7dhoHMvCqgnPao';

	// Act + Assert:
	return runMetadataDbTest(
		dbMetadata(),
		db => db.binDataByMetalId(metalId).catch(error => error),
		error => expect(error.message).to.equal(`could not get first chunk, it may mistake the metal ID: ${metalId}`)
	);
});
```